### PR TITLE
Manage Password History

### PR DESCRIPTION
**Pull request by hlbraddock**

### CODE REVIEW

**Comment by hlbraddock** [comment]
@frett This PR is not yet complete, as I'd like to add a few more tests. But you are free to take a look at your leisure.

File: api/src/main/java/org/ccci/idm/user/dao/ldap/AbstractLdapUserDao.java [hunk 1]
```diff
@@ -71,6 +72,7 @@
             .put(Attr.CONTACT, ImmutableSet.of(LDAP_ATTR_TELEPHONE, LDAP_ATTR_OBJECTCLASS))
             .put(Attr.CRU_PREFERRED_NAME, ImmutableSet.of(LDAP_ATTR_CRU_PREFERRED_NAME, LDAP_ATTR_OBJECTCLASS))
             .put(Attr.CRU_PROXY_ADDRESSES, ImmutableSet.of(LDAP_ATTR_CRU_PROXY_ADDRESSES, LDAP_ATTR_OBJECTCLASS))
+            .put(Attr.CRU_PASSWORD_HISTORY, ImmutableSet.of(LDAP_ATTR_CRU_PASSWORD_HISTORY, LDAP_ATTR_OBJECTCLASS))
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/dao/ldap/AbstractLdapUserDao.java]: I would be fine with including `LDAP_ATTR_CRU_PASSWORD_HISTORY` in the `Attr.PASSWORD` set, that way all existing code will automatically save the password history when changing the password.

**Comment by hlbraddock** [comment]
@frett I've incorporated your suggestions and expanded on password history testing. Provide final review and any other comments you may have.

File: api/src/main/java/org/ccci/idm/user/dao/ldap/AbstractLdapUserDao.java (hunk 1 above)
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/dao/ldap/AbstractLdapUserDao.java]: *bump*

File: api/src/main/java/org/ccci/idm/user/DefaultUserManager.java [hunk 2]
```diff
@@ -155,6 +164,11 @@ public void updateUser(final User user, final User.Attr... attrs) throws DaoExce
         // update the user object
         this.userDao.update(original, user, attrs);
 
+        // add password to history (if you have password and caller intends to set)
+        if(StringUtils.hasText(user.getPassword()) && Iterables.contains(Arrays.asList(attrs), User.Attr.PASSWORD)) {
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/DefaultUserManager.java]: something I just tried out that makes the code more readable:

`FluentIterable.of(attrs).contains(User.Attr.PASSWORD)`

this isn't blocking, just a more readable way to check for the Password attr set

File: api/src/main/java/org/ccci/idm/user/util/PasswordHistoryManager.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    private static final Splitter HASH_TIME_STAMP_SPLITTER = Splitter.on(HASH_TIME_STAMP_DELIMITER);
+
+    private static final Ordering<String> CHRONOLOGICAL_ORDERING = new Ordering<String>() {
+        @Override
+        public int compare(String first, String second) {
+            try {
+                DateTime firstDateTime = new DateTime(Iterables.getLast(HASH_TIME_STAMP_SPLITTER.split(first)));
+                DateTime secondDateTime = new DateTime(Iterables.getLast(HASH_TIME_STAMP_SPLITTER.split(second)));
+
+                if (firstDateTime.isBefore(secondDateTime)) {
```
Comment (frett) [inline api/src/main/java/org/ccci/idm/user/util/PasswordHistoryManager.java]: a comment that was lost elsewhere, this if block could be simplified to:

```
return secondDateTime.compareTo(firstDateTime);
```

**Comment by frett** [comment]
just a couple minor things that got lost in the previous conversation, other than that it looks good.

I already started working on the UserManagerListener to observe password history for Secure Accounts

File: api/src/main/java/org/ccci/idm/user/util/PasswordHistoryManager.java (hunk 3 above)
Comment (hlbraddock) [inline api/src/main/java/org/ccci/idm/user/util/PasswordHistoryManager.java]: Great. I was looking for that. Done.

File: api/src/main/java/org/ccci/idm/user/DefaultUserManager.java (hunk 2 above)
Comment (hlbraddock) [inline api/src/main/java/org/ccci/idm/user/DefaultUserManager.java]: Done

**Comment by hlbraddock** [comment]
Awww shucks, I was hoping to do that. I know you're wanting to knock this out.

**Comment by hlbraddock** [comment]
I'm just embellishing testing of password history through `LdaptiveUserDao` as per your suggestion to include `LDAP_ATTR_CRU_PASSWORD_HISTORY` in the `Attr.PASSWORD` set.

**Comment by hlbraddock** [comment]
I fixed an issue in `DefaultUserManager` such that I add password to history prior to persisting.

In addition, unless we expect that caller to specify password history as a user attribute (which we obviously wouldn't) or we add the attribute to the list ourselves (which is not what we'd want), it is actually required that we include `LDAP_ATTR_CRU_PASSWORD_HISTORY` in the `Attr.PASSWORD` mapped set.

I've modified `testUpdateUser()` of `AbstractDefaultUserManagerIT` to ensure password history is properly persisted when password is modified.

**Comment by hlbraddock** [comment]
@frett refactored not to implement action at a distance and to allow for passing unmodifiable collection.

**Comment by frett** [comment]
looks good +1

**Pull request merged**